File: 110. Balanced Binary Tree/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function(root) {
  return dfs(root) !== -1;
};

const dfs = function(root) {
  if (!root) return 0;

  const leftDepth = dfs(root.left);
  if (leftDepth === -1) return -1;
  const rightDepth = dfs(root.right);
  if (rightDepth === -1) return -1;
  if (Math.abs(leftDepth - rightDepth) > 1) return -1;
  return Math.max(leftDepth, rightDepth) + 1;
};

File: 110. Balanced Binary Tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        result = True

        def counter(node):
            nonlocal result
            if not node:
                return 0
            dl = counter(node.left)
            if not result:
                return 0
            dr = counter(node.right)
            if abs(dl - dr) > 1:
                result = False
            return max(dl, dr) + 1
        counter(root)
        return result
